packages/common/solana/utils.ts

```diff
@@ -136,17 +135,97 @@ export async function signAndExecuteTransaction(
+ * but is not in `commitment` state and so it's not considered executed by the `sendAndConfirmRawTransaction` method,
+ * and it raises an expiry error even though transaction may be executed soon.
+ * - so we add additional 50 blocks for checks to account for such issues;
+ * - also, we check for SignatureStatus one last time as it could be that websocket was slow to respond.
+ * @param connection - Solana client connection
+ * @param tx - Transaction instance
+ * @param hash - blockhash information, the same hash should be used in the Transaction
+ * @returns Transaction signature
+ */
+export async function executeTransaction(
+  connection: Connection,
+  tx: Transaction,
+  hash: BlockhashWithExpiryBlockHeight,
+): Promise<string> {
+  const rawTx = tx.serialize();
+
+  if (!hash.lastValidBlockHeight || !tx.signature || !hash.blockhash) throw Error("Error with transaction parameters.");
 
+  const signature = bs58.encode(tx.signature);
   const confirmationStrategy: BlockheightBasedTransactionConfirmationStrategy = {
     lastValidBlockHeight: hash.lastValidBlockHeight + 50,
-    signature: bs58.encode(signedTx.signature),
+    signature,
     blockhash: hash.blockhash,
   };
-  return sendAndConfirmRawTransaction(connection, rawTx, confirmationStrategy);
+  try {
+    return await sendAndConfirmRawTransaction(connection, rawTx, confirmationStrategy);
+  } catch (e) {
+    // If BlockHeight expired, we will check tx status one last time to make sure
+    if (e instanceof TransactionExpiredBlockheightExceededError) {
+      await sleep(1000);
+      const value = await confirmAndEnsureTransaction(connection, signature);
+      if (!value) {
+        throw e;
+      }
+      return signature;
+    }
+    throw e;
+  }
+}
+
+/**
+ * Confirms and validates transaction success once
+ * @param connection - Solana client connection
+ * @param signature - Transaction signature
+ * @returns Transaction Status
+ */
+export async function confirmAndEnsureTransaction(
+  connection: Connection,
+  signature: string,
+): Promise<SignatureStatus | null> {
+  const response = await connection.getSignatureStatus(signature);
+  if (!response) {
+    return null;
+  }
+  const { value } = response;
+  if (!value) {
+    return null;
+  }
+  if (value.err) {
+    // That's how solana-web3js does it, `err` here is an object that won't really be handled
+    throw new Error(`Raw transaction ${signature} failed (${JSON.stringify({ err: value.err })})`);
+  }
+  switch (connection.commitment) {
+    case "confirmed":
+    case "single":
+    case "singleGossip": {
+      if (value.confirmationStatus === "processed") {
+        return null;
+      }
+      break;
+    }
+    case "finalized":
+    case "max":
+    case "root": {
+      if (value.confirmationStatus === "processed" || value.confirmationStatus === "confirmed") {
+        return null;
+      }
+      break;
+    }
+    // exhaust enums to ensure full coverage
+    case "processed":
+    case "recent":
```

Comment:
what behavior do you expect here? Just a fallthrough to the final `return`? 
For me personally, this code style is harder to read, I'd suggest to be more specific here and put the return statement inside the switch


nit: btw, `// exhaust enums to ensure full coverage`
to achieve this you could use
```
    default: {
      const defaultValue: never = connection.commitment;
      return whatever;
    }
```
The gotcha here is `never` type, so if any meaningful value falls through to the `default` case, TS will catch it.